File: example/src/main/java/com/kekwy/cfn/example/Resource.java
package com.kekwy.cfn.example;

import lombok.AllArgsConstructor;
import lombok.Data;

import java.util.Collection;

@Data
@AllArgsConstructor
public class Resource {
    private int cpu;
    private int memory;
    private Collection<String> tags;

    @SuppressWarnings("BooleanMethodIsAlwaysInverted")
    public boolean check(Resource requiredResource) {
        return cpu >= requiredResource.cpu && memory >= requiredResource.memory
                && tags.containsAll(requiredResource.tags);
    }

    public void allocate(Resource requiredResource) {
        // TODO: tag 所表示的设备是否为独占，或也需要对其负载进行量化
        cpu -= requiredResource.cpu;
        memory -= requiredResource.memory;
    }


    public void recycle(Resource recycledResource) {
        cpu += recycledResource.cpu;
        memory += recycledResource.memory;
    }
}

File: models/resource.go
package models

import "github.com/kekwy/cfn/messages"

type Resource struct {
	CPU    *CPU
	GPU    *GPU
	Memory *Memory
	Tags   []string
}

func (resource *Resource) toMessage() messages.Resource {
	return messages.Resource{
		CPU:    resource.CPU.toMessage(),
		GPU:    resource.GPU.toMessage(),
		Memory: resource.Memory.toMessage(),
		Tags:   resource.Tags,
	}
}

type CPU struct {
	Cores int32
}

func (cpu *CPU) toMessage() *messages.CPU {
	return &messages.CPU{
		Cores: cpu.Cores,
	}
}

type GPU struct {
	Cores int32
	Model string
}

func (gpu *GPU) toMessage() *messages.GPU {
	return &messages.GPU{
		Cores: gpu.Cores,
		Model: gpu.Model,
	}
}

type Memory struct {
	Capacity int32
}

func (memory *Memory) toMessage() *messages.Memory {
	return &messages.Memory{
		Capacity: memory.Capacity,
	}
}
